pkg/stores.go

```diff
@@ -1,41 +1,76 @@
-	Remove(T)
-	View() ([]T, error)
+const (
+	NotifyOpNil NotifyOp = iota
+	// NotifyOpEvict is a notification that a result has been evicted from the store after TTL has passed
+	NotifyOpEvict
+	// NotifyOpRemove is a notification that a result has been removed from the store
+	NotifyOpRemove
+)
+
+// ResultStore stores check results and allows for querying and removing result
+type ResultStore interface {
+	Add(...CheckResult)
+	Remove(...string)
+	View(...ViewOpt) ([]CheckResult, error)
+	Notifications() <-chan Notification
 }
 
-type resultStore[T any] struct {
-	data map[string]T
+// Notification is a struct that will be sent by the ResultStore upon certain events happening
+type Notification struct {
+	Op   NotifyOp
+	Data CheckResult
 }
 
-func NewResultStore[T any]() *resultStore[T] {
-	return &resultStore[T]{
-		data: make(map[string]T),
+// Filter is a function that filters check results from a ResultStore
+type ResultFilter func(CheckResult) bool
+
+// Comparator is a function that is used for ordering of results from a ResultStore.
+// It should return true if the first result should be ordered before the second result.
+type ResultComparator func(i, j CheckResult) bool
+
+// ViewOpts is a set of options that can be passed to the View method of a ResultStore
+type viewOpts struct {
+	filters     []ResultFilter
+	comparators []ResultComparator
+	limit       int
+}
+
+type ViewOpts []ViewOpt
+
+// Apply applies the ViewOpts to a viewOpts and destructs it into filters, comparators and limit.
+// TODO: TBD if we want to keep this or just use the viewOpts directly or as arguments in the View method.
+func (vo ViewOpts) Apply() ([]ResultFilter, []ResultComparator, int) {
+	opts := &viewOpts{}
+	for _, opt := range vo {
+		opt(opts)
 	}
+	return opts.filters, opts.comparators, opts.limit
 }
 
-func (s *resultStore[T]) Add(results ...T) {
-	for _, result := range results {
-		key := fmt.Sprintf("%v", result)
-		s.data[key] = result
+// ViewOpt is an option that can be passed to the View method of a ResultStore
+type ViewOpt func(*viewOpts)
+
+// WithResultFilters enables to specify filters that will be applied to the results
+func WithFilter(filters ...ResultFilter) ViewOpt {
+	return func(opts *viewOpts) {
+		opts.filters = append(opts.filters, filters...)
 	}
 }
 
-func (s *resultStore[T]) Remove(results ...T) {
-	for _, result := range results {
-		key := fmt.Sprintf("%v", result)
-		delete(s.data, key)
+// WithOrder enables to specify the order in which the results should be returned
+func WithOrder(comparators ...ResultComparator) ViewOpt {
+	return func(opts *viewOpts) {
+		opts.comparators = append(opts.comparators, comparators...)
 	}
 }
 
-func (s *resultStore[T]) View() ([]T, error) {
-	var result []T
-	for _, r := range s.data {
-		result = append(result, r)
+// WithOrderView enables to limit the amount of results that should be returned
+func WithLimit(limit int) ViewOpt {
+	return func(opts *viewOpts) {
+		opts.limit = limit
```

Comment:
Do all of the interfaces and types declared in this file need to be exported? Or can some of them be unexported? 🏄

Reply:
We can maybe place some in the `./resultstore` package